relayer/relayer/src/trader/mod.rs

```diff
@@ -0,0 +1,303 @@
+
+#[derive(Debug, Error)]
+#[error(transparent)]
+#[non_exhaustive]
+pub enum TraderError {
+    #[error("Error when parsing ethereum address {0}")]
+    FromHex(#[from] rustc_hex::FromHexError),
+
+    #[error("AlephClient error {0}")]
+    AlephClient(#[from] anyhow::Error),
+
+    #[error("Azero contract error {0}")]
+    AzeroContract(#[from] AzeroContractError),
+
+    #[error("Broadcast receive error {0}")]
+    BroadcastReceive(#[from] broadcast::error::RecvError),
+
+    #[error("Missing required arg {0}")]
+    MissingRequired(String),
+
+    #[error("Address is not an AccountId {0}")]
+    NotAccountId(String),
+
+    #[error("Trader has exited unexpectedly - this should never happen")]
+    TraderExited,
+}
+
+#[derive(Copy, Clone)]
+pub struct Trader;
+
+impl Trader {
+    pub async fn run(
+        config: Arc<Config>,
+        azero_signed_connection: Arc<AzeroConnectionWithSigner>,
+        eth_signed_connection: Arc<SignedEthConnection>,
+        mut circuit_breaker_receiver: broadcast::Receiver<CircuitBreakerEvent>,
+    ) -> Result<CircuitBreakerEvent, TraderError> {
+        let Config {
+            azero_contract_metadata,
+            azero_contract_address,
+            azero_ref_time_limit,
+            azero_proof_size_limit,
+            router_address,
+            router_metadata,
+            azero_wrapped_azero_address,
+            azero_ether_address,
+            azero_ether_metadata,
+            ..
+        } = &*config;
+
+        let most_azero = MostInstance::new(
+            azero_contract_address,
+            azero_contract_metadata,
+            *azero_ref_time_limit,
+            *azero_proof_size_limit,
+        )?;
+
+        let router = RouterInstance::new(
+            &router_address
+                .clone()
+                .ok_or(TraderError::MissingRequired("router_address".to_owned()))?,
+            router_metadata,
+            *azero_ref_time_limit,
+            *azero_proof_size_limit,
+        )?;
+
+        let azero_ether_address =
+            azero_ether_address
+                .clone()
+                .ok_or(TraderError::MissingRequired(
+                    "azero_ether_address".to_owned(),
+                ))?;
+
+        let azero_ether = AzeroEtherInstance::new(
+            &azero_ether_address,
+            azero_ether_metadata,
+            *azero_ref_time_limit,
+            *azero_proof_size_limit,
+        )?;
+
+        let wrapped_azero_address =
+            AccountId::from_str(&azero_wrapped_azero_address.clone().ok_or(
+                TraderError::MissingRequired("azero_wrapped_azero_address".to_owned()),
+            )?)
+            .map_err(|err| TraderError::NotAccountId(err.to_owned()))?;
+
+        let whoami_azero = azero_signed_connection.account_id();
+        let whoami_eth = eth_signed_connection.address().to_string();
+
+        let swap_path = [wrapped_azero_address.clone(), azero_ether.address.clone()];
```

Comment:
Suggestion: Perhaps you could make the struct do some work after all and change these "helper variables" into helper functions or initialize them in `new` and then have a slimmed-down `run`?